File: src/lib/stats/aggregator.ts
/**
 * src/lib/stats/aggregator.ts — Agrégation pure items × mentions × episodes
 * pour les pages stats publiques.
 *
 * Cf. ADR 0047. Pures fonctions, idempotentes (tri stable, pas d'aléa).
 *
 * Filtrage : on exclut systématiquement les mentions `status='discarded'`
 * (cohérent avec `gallery/aggregate.ts::publicMentions`). Les citations
 * (`kind='citation'`) sont incluses car la stat est une vue d'ensemble.
 * Les œuvres d'invité·es (`guestWork`) restent `kind='reco'` et sont donc
 * comptées comme des recommandations normales ici (décision produit CR
 * Story 4 : seule la page épisode les présente à part).
 *
 * Issues fixées :
 *  - H26-2 : tri via `frSortKey` (NFKD) déterministe au lieu de
 *    `localeCompare` (ICU runtime-dependent).
 *  - M26-6 : `coerceMonth` borne l'année à [1900, 2100] (filtre les dates
 *    aberrantes type 0001 ou 9999).
 *  - M26-19 : dédoublonnement des slugs (collisions accents).
 *  - L26-27 : `computeTypeDistribution` tri secondaire `count DESC` puis
 *    alpha.
 *  - R-P3-30 : `monthlyEpisodes` remplit les mois manquants avec `count=0`
 *    entre min et max pour des charts continus.
 */
import { frSortKey, slugify, uniqueSlug } from './slug';
import type {
  GlobalCounts,
  MonthlyBucket,
  StatsSnapshot,
  TopGuest,
  TopWork,
} from './types';
import { STATS_SCHEMA_VERSION } from './types';

const HIDDEN_STATUS = new Set(['discarded']);

/**
 * Borne inférieure de l'année acceptée par `coerceMonth`.
 *
 * Pourquoi : filtre les dates aberrantes type `0001-01-01` qu'on rencontre
 * parfois dans les flux RSS mal formés (M26-6). Exporté pour réutilisation
 * dans les tests et les scripts Python miroirs (`tools/stats/aggregator.py`).
 */
export const MONTH_MIN_YEAR = 1900;

/**
 * Borne supérieure de l'année acceptée par `coerceMonth`.
 *
 * Pourquoi : filtre les dates aberrantes type `9999-12-31` (sentinelles RSS,
 * timestamps mal convertis). Exporté pour réutilisation dans les tests et
 * les scripts Python miroirs.
 */
export const MONTH_MAX_YEAR = 2100;

/**
 * Fenêtre maximale (en mois) du chart « épisodes/mois » (F-CRIT-9).
 *
 * Pourquoi : `computeMonthlyEpisodes` remplit les mois manquants entre min
 * et max pour des charts continus (R-P3-30). Sans cap, une mention parasite
 * avec une date du futur (ou des archives 1990) produirait des milliers de
 * buckets vides et exploserait le HTML/SVG. On garde 5 ans glissants à
 * partir du dernier mois observé.
 */
export const MONTHLY_WINDOW_MAX = 60;

// --- Inputs (formes minimales — duck-typing pour testabilité) ---------------

export interface ItemLike {
  id: string;
  title: string;
  types: readonly string[];
}

export interface MentionLike {
  itemId: string;
  recommendedBy?: string | null;
  status?: 'draft' | 'validated' | 'discarded';
  sourceRef: { sourceId: string };
}

export interface EpisodeLike {
  sourceId: string;
  date?: Date | string | null;
}

export interface SourceLike {
  id: string;
  hosts?: readonly string[];
}

export interface AggregateOptions {
  topGuestsLimit?: number;
  topWorksLimit?: number;
  /** Restreint l'agrégation à une source unique (pour `/[source]/stats`). */
  sourceId?: string;
  /** ISO timestamp à inscrire dans `generatedAt`. Défaut : `new Date().toISOString()`. */
  generatedAt?: string;
}

// --- Helpers ----------------------------------------------------------------

/** Garde uniquement les mentions publiques. */
export function publicMentions(mentions: readonly MentionLike[]): MentionLike[] {
  return mentions.filter((m) => !HIDDEN_STATUS.has(m.status ?? 'draft'));
}

function buildHostSet(sources: readonly SourceLike[]): Set<string> {
  const hosts = new Set<string>();
  for (const s of sources) {
    for (const h of s.hosts ?? []) {
      const k = h.trim().toLowerCase();
      if (k) hosts.add(k);
    }
  }
  return hosts;
}

/** Décale un mois ISO `YYYY-MM` de `delta` mois (peut être négatif). */
function shiftMonth(month: string, delta: number): string {
  const [y, m] = month.split('-').map((n) => Number(n));
  // Index 0-based pour rester dans l'arithmétique modulo 12 propre.
  const total = y * 12 + (m - 1) + delta;
  const ny = Math.floor(total / 12);
  const nm = (total % 12 + 12) % 12 + 1;
  return `${ny}-${String(nm).padStart(2, '0')}`;
}

function coerceMonth(date: Date | string | null | undefined): string | null {
  if (!date) return null;
  const d = date instanceof Date ? date : new Date(date);
  if (Number.isNaN(d.getTime())) return null;
  const y = d.getUTCFullYear();
  if (y < MONTH_MIN_YEAR || y > MONTH_MAX_YEAR) return null;
  const m = String(d.getUTCMonth() + 1).padStart(2, '0');
  return `${y}-${m}`;
}

// --- Counts ------------------------------------------------------------------

/**
 * Calcule les compteurs globaux à partir des collections déjà filtrées.
 * Le filtrage par source doit avoir été appliqué en amont si nécessaire.
 */
export function computeGlobalCounts(args: {
  sources: readonly SourceLike[];
  episodes: readonly EpisodeLike[];
  mentions: readonly MentionLike[];
  items: readonly ItemLike[];
}): GlobalCounts {
  const pub = publicMentions(args.mentions);
  const hosts = buildHostSet(args.sources);
  const mentionedItemIds = new Set<string>();
  const guests = new Set<string>();
  for (const m of pub) {
    mentionedItemIds.add(m.itemId);
    const by = (m.recommendedBy ?? '').trim();
    if (!by) continue;
    const key = by.toLowerCase();
    if (hosts.has(key)) continue;
    guests.add(key);
  }
  // Une œuvre n'est comptée que si elle est effectivement mentionnée et présente
  // dans le catalogue d'items (cohérent avec les galeries).
  const itemIds = new Set(args.items.map((i) => i.id));
  let uniqueWorks = 0;
  for (const id of mentionedItemIds) if (itemIds.has(id)) uniqueWorks += 1;

  return {
    podcastsCount: args.sources.length,
    episodesCount: args.episodes.length,
    recommendationsCount: pub.length,
    uniqueWorksCount: uniqueWorks,
    uniqueGuestsCount: guests.size,
  };
}

// --- Top guests --------------------------------------------------------------

export function computeTopGuests(
  mentions: readonly MentionLike[],
  sources: readonly SourceLike[],
  limit = 10,
): TopGuest[] {
  const hosts = buildHostSet(sources);
  /**
   * F-H-14 : par clé lower-case, on garde un Counter des variantes du nom
   * croisées dans les mentions. À la sortie on choisit la forme la plus
   * fréquente (tie → on préfère la version avec au moins une majuscule —
   * « Alice » vs « alice » — qui reflète une saisie soignée).
   */
  const counts = new Map<
    string,
    { variants: Map<string, number>; count: number }
  >();
  for (const m of publicMentions(mentions)) {
    const raw = (m.recommendedBy ?? '').trim();
    if (!raw) continue;
    const key = raw.toLowerCase();
    if (hosts.has(key)) continue;
    let cur = counts.get(key);
    if (!cur) {
      cur = { variants: new Map(), count: 0 };
      counts.set(key, cur);
    }
    cur.count += 1;
    cur.variants.set(raw, (cur.variants.get(raw) ?? 0) + 1);
  }
  function pickName(variants: Map<string, number>): string {
    let bestName = '';
    let bestCount = -1;
    let bestHasUpper = false;
    for (const [name, n] of variants) {
      const hasUpper = name !== name.toLowerCase();
      if (n > bestCount) {
        bestName = name;
        bestCount = n;
        bestHasUpper = hasUpper;
      } else if (n === bestCount && hasUpper && !bestHasUpper) {
        // Tie → préfère la version capitalisée.
        bestName = name;
        bestHasUpper = true;
      }
    }
    return bestName;
  }
  const resolved = [...counts.values()].map(({ variants, count }) => ({
    name: pickName(variants),
    count,
  }));
  const entries = resolved.sort((a, b) => {
    if (b.count !== a.count) return b.count - a.count;
    const ka = frSortKey(a.name);
    const kb = frSortKey(b.name);
    return ka < kb ? -1 : ka > kb ? 1 : 0;
  });
  const used = new Set<string>();
  return entries.slice(0, limit).map(({ name, count }) => ({
    name,
    slug: uniqueSlug(name, used),
    count,
  }));
}

// --- Top works ---------------------------------------------------------------

export function computeTopWorks(
  items: readonly ItemLike[],
  mentions: readonly MentionLike[],
  limit = 10,
): TopWork[] {
  const counts = new Map<string, number>();
  for (const m of publicMentions(mentions)) {
    counts.set(m.itemId, (counts.get(m.itemId) ?? 0) + 1);
  }
  const itemById = new Map(items.map((i) => [i.id, i]));
  const out: TopWork[] = [];
  for (const [id, n] of counts) {
    const it = itemById.get(id);
    if (!it) continue;
    out.push({
      id,
      title: it.title,
      type: it.types[0] ?? 'autre',
      mentionsCount: n,
    });
  }
  out.sort((a, b) => {
    if (b.mentionsCount !== a.mentionsCount) return b.mentionsCount - a.mentionsCount;
    const ka = frSortKey(a.title);
    const kb = frSortKey(b.title);
    return ka < kb ? -1 : ka > kb ? 1 : 0;
  });
  return out.slice(0, limit);
}

// --- Type distribution -------------------------------------------------------

/**
 * Distribution `type → count` (1 item par type primaire = `types[0]`).
 * Une œuvre n'apparaît qu'une fois (clé `id`), agrégée sur son type principal.
 *
 * Tri : `count DESC` puis alpha (L26-27).
 */
export function computeTypeDistribution(
  items: readonly ItemLike[],
  mentions: readonly MentionLike[],
): Record<string, number> {
  const mentionedIds = new Set<string>();
  for (const m of publicMentions(mentions)) mentionedIds.add(m.itemId);
  const dist: Record<string, number> = {};
  for (const it of items) {
    if (!mentionedIds.has(it.id)) continue;
    const t = it.types[0] ?? 'autre';
    dist[t] = (dist[t] ?? 0) + 1;
  }
  const sortedKeys = Object.keys(dist).sort((a, b) => {
    if (dist[b] !== dist[a]) return dist[b] - dist[a];
    const ka = frSortKey(a);
    const kb = frSortKey(b);
    return ka < kb ? -1 : ka > kb ? 1 : 0;
  });
  const sorted: Record<string, number> = {};
  for (const k of sortedKeys) sorted[k] = dist[k];
  return sorted;
}

// --- Monthly episodes --------------------------------------------------------

/**
 * Agrège les épisodes par mois ISO `YYYY-MM` et **remplit les mois
 * manquants entre min et max avec `count=0`** (R-P3-30) pour des charts
 * continus (sinon les bar charts laissent des trous trompeurs).
 *
 * F-CRIT-9 : la fenêtre est **cappée à `MONTHLY_WINDOW_MAX` mois** (60 par
 * défaut) à partir du dernier mois observé. Sans cap, une seule date
 * aberrante (futur 2099, archive 1995…) produirait des centaines de buckets
 * vides et exploserait le SVG / le HTML.
 */
export function computeMonthlyEpisodes(
  episodes: readonly EpisodeLike[],
): MonthlyBucket[] {
  const counts = new Map<string, number>();
  for (const e of episodes) {
    const m = coerceMonth(e.date);
    if (!m) continue;
    counts.set(m, (counts.get(m) ?? 0) + 1);
  }
  if (counts.size === 0) return [];
  const present = [...counts.keys()].sort();
  let first = present[0];
  const last = present[present.length - 1];
  // F-CRIT-9 — Cap la fenêtre à `MONTHLY_WINDOW_MAX` mois.
  const minAllowed = shiftMonth(last, -(MONTHLY_WINDOW_MAX - 1));
  if (first < minAllowed) first = minAllowed;
  const out: MonthlyBucket[] = [];
  let [y, mo] = first.split('-').map((n) => Number(n));
  const [yLast, moLast] = last.split('-').map((n) => Number(n));
  while (y < yLast || (y === yLast && mo <= moLast)) {
    const key = `${y}-${String(mo).padStart(2, '0')}`;
    out.push({ month: key, count: counts.get(key) ?? 0 });
    mo += 1;
    if (mo > 12) {
      mo = 1;
      y += 1;
    }
  }
  return out;
}

// --- Per-source --------------------------------------------------------------

export function computePerSource(args: {
  sources: readonly SourceLike[];
  episodes: readonly EpisodeLike[];
  mentions: readonly MentionLike[];
  items: readonly ItemLike[];
}): Record<string, GlobalCounts> {
  const out: Record<string, GlobalCounts> = {};
  for (const s of args.sources) {
    const eps = args.episodes.filter((e) => e.sourceId === s.id);
    const mens = args.mentions.filter((m) => m.sourceRef.sourceId === s.id);
    const mentionedIds = new Set(mens.map((m) => m.itemId));
    const its = args.items.filter((i) => mentionedIds.has(i.id));
    out[s.id] = computeGlobalCounts({
      sources: [s],
      episodes: eps,
      mentions: mens,
      items: its,
    });
  }
  return out;
}

// --- Build reproductibilité --------------------------------------------------

/**
 * Résout `generatedAt` depuis les variables d'environnement de build :
 *  1. `SOURCE_DATE_EPOCH` — secondes Unix (convention reproducible-builds.org).
 *  2. `RECO_BUILD_TIMESTAMP` — ISO 8601 brut (override projet).
 *
 * Retourne `null` si aucune var n'est positionnée / parseable, charge à
 * l'appelant de fallback sur `new Date().toISOString()`.
 */
function resolveGeneratedAtFromEnv(): string | null {
  const epoch = process.env.SOURCE_DATE_EPOCH;
  if (epoch && /^\d+$/.test(epoch)) {
    const sec = Number(epoch);
    if (Number.isFinite(sec)) {
      const iso = new Date(sec * 1000).toISOString();
      // Normalise sur la forme stricte attendue par `statsSnapshotSchema`
      // (avec millisecondes — `toISOString` la produit déjà).
      return iso;
    }
  }
  const raw = process.env.RECO_BUILD_TIMESTAMP;
  if (raw && raw.trim()) return raw.trim();
  return null;
}

// --- Façade ------------------------------------------------------------------

export function buildStatsSnapshot(args: {
  sources: readonly SourceLike[];
  episodes: readonly EpisodeLike[];
  mentions: readonly MentionLike[];
  items: readonly ItemLike[];
  options?: AggregateOptions;
}): StatsSnapshot {
  const opts = args.options ?? {};
  let sources = args.sources;
  let episodes = args.episodes;
  let mentions = args.mentions;
  let items = args.items;
  if (opts.sourceId) {
    sources = sources.filter((s) => s.id === opts.sourceId);
    episodes = episodes.filter((e) => e.sourceId === opts.sourceId);
    mentions = mentions.filter((m) => m.sourceRef.sourceId === opts.sourceId);
    const mentionedIds = new Set(mentions.map((m) => m.itemId));
    items = items.filter((i) => mentionedIds.has(i.id));
  }

  // F-CRIT-4 — build reproductible : priorité aux env vars `SOURCE_DATE_EPOCH`
  // (convention reproducible-builds.org) puis `RECO_BUILD_TIMESTAMP`, puis
  // fallback temps réel. Permet à un re-build de produire un snapshot
  // bit-identique aux fins de cache CDN / signatures.
  const generatedAt =
    opts.generatedAt ??
    resolveGeneratedAtFromEnv() ??
    new Date().toISOString();

  return {
    schemaVersion: STATS_SCHEMA_VERSION,
    generatedAt,
    global: computeGlobalCounts({ sources, episodes, mentions, items }),
    perSource: computePerSource({ sources, episodes, mentions, items }),
    topGuests: computeTopGuests(mentions, sources, opts.topGuestsLimit ?? 50),
    topWorks: computeTopWorks(items, mentions, opts.topWorksLimit ?? 50),
    typeDistribution: computeTypeDistribution(items, mentions),
    monthlyEpisodes: computeMonthlyEpisodes(episodes),
  };
}

// Re-export pour rétrocompat des tests existants qui importent slugify
// depuis `./slug` (ne pas casser l'API publique du package).
export { slugify };

File: tools/stats/aggregator.py
"""Agrégation des stats publiques (pur Python, idempotent — ADR 0047).

Pendant du module TypeScript `src/lib/stats/aggregator.ts`. Mêmes invariants :

- exclusion des mentions `status='discarded'` ;
- exclusion des hôtes du podcast dans le décompte des invités uniques ;
- tri stable (locale FR pour les noms, ISO pour les mois) ;
- pas d'aléa, pas d'I/O — la lecture des collections est faite par
  `tools.build_stats`.

B-MED-12 — **Python 3.11+ requis** : on utilise ``datetime.fromisoformat``
qui n'accepte le suffixe ``Z`` natif qu'à partir de 3.11. Cette
dépendance est tracée dans ``pyproject.toml`` (``requires-python``).
"""
from __future__ import annotations

import re
import unicodedata
from collections.abc import Iterable, Mapping
from datetime import datetime, timezone
from typing import Any

from .models import (
    GlobalCounts,
    MonthlyBucket,
    StatsSnapshot,
    TopGuest,
    TopWork,
)

_HIDDEN_STATUSES = frozenset({"discarded"})
_RE_NONALNUM = re.compile(r"[^a-z0-9]+")


# --- Helpers ----------------------------------------------------------------


def _slugify(value: str) -> str:
    """Slug ASCII minuscule cohérent avec `tools.common.slugify`."""
    if not value:
        return "x"
    norm = unicodedata.normalize("NFKD", value)
    norm = norm.encode("ascii", "ignore").decode("ascii").lower()
    norm = _RE_NONALNUM.sub("-", norm).strip("-")
    return norm or "x"


def _is_public(mention: Mapping[str, Any]) -> bool:
    return mention.get("status", "draft") not in _HIDDEN_STATUSES


def public_mentions(mentions: Iterable[Mapping[str, Any]]) -> list[dict[str, Any]]:
    """Garde les mentions publiques (status != discarded)."""
    return [dict(m) for m in mentions if _is_public(m)]


def _build_hosts(sources: Iterable[Mapping[str, Any]]) -> set[str]:
    out: set[str] = set()
    for s in sources:
        for h in s.get("hosts") or []:
            k = str(h).strip().lower()
            if k:
                out.add(k)
    return out


def _month_key(value: Any) -> str | None:
    """Convertit une date (`datetime`, ISO string, None) en `YYYY-MM`."""
    if value is None or value == "":
        return None
    dt: datetime | None = None
    if isinstance(value, datetime):
        dt = value
    elif isinstance(value, str):
        try:
            # `fromisoformat` accepte les ISO simples ; on retire un suffixe Z.
            dt = datetime.fromisoformat(value.replace("Z", "+00:00"))
        except ValueError:
            return None
    if dt is None:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    dt = dt.astimezone(timezone.utc)
    return f"{dt.year:04d}-{dt.month:02d}"


def _fr_sort_key(s: str) -> str:
    """Clé de tri locale FR insensible aux accents/casses."""
    norm = unicodedata.normalize("NFKD", s)
    return "".join(c for c in norm if not unicodedata.combining(c)).lower()


# --- Computes ---------------------------------------------------------------


def compute_global_counts(
    *,
    sources: list[Mapping[str, Any]],
    episodes: list[Mapping[str, Any]],
    mentions: list[Mapping[str, Any]],
    items: list[Mapping[str, Any]],
) -> GlobalCounts:
    pub = public_mentions(mentions)
    hosts = _build_hosts(sources)
    mentioned_ids: set[str] = set()
    guests: set[str] = set()
    for m in pub:
        mentioned_ids.add(str(m["itemId"]))
        # B-MED-14 — guard `isinstance(..., str)` : un dataset legacy peut
        # avoir un `recommendedBy` typé int ou None. On veut un fail mou
        # (l'entrée est ignorée), pas un AttributeError sur `.strip()`.
        rec = m.get("recommendedBy")
        if not isinstance(rec, str):
            continue
        raw = rec.strip()
        if not raw:
            continue
        # B-MED-13 — `key = raw.lower()` : dédoublonnement insensible à la
        # casse (« Doe » et « doe » comptent comme un seul invité).
        # `.lower()` est suffisant ici (entrées libellées en latin) ;
        # le tri d'affichage utilise `_fr_sort_key` qui retire aussi les
        # accents pour le rang final.
        key = raw.lower()
        if key in hosts:
            continue
        guests.add(key)
    item_ids = {str(i["id"]) for i in items}
    unique_works = sum(1 for x in mentioned_ids if x in item_ids)
    return GlobalCounts(
        podcastsCount=len(sources),
        episodesCount=len(episodes),
        recommendationsCount=len(pub),
        uniqueWorksCount=unique_works,
        uniqueGuestsCount=len(guests),
    )


def compute_top_guests(
    mentions: list[Mapping[str, Any]],
    sources: list[Mapping[str, Any]],
    limit: int = 10,
) -> list[TopGuest]:
    hosts = _build_hosts(sources)
    counts: dict[str, dict[str, Any]] = {}
    for m in public_mentions(mentions):
        # B-MED-14 — même garde isinstance() que dans compute_global_counts.
        rec = m.get("recommendedBy")
        if not isinstance(rec, str):
            continue
        raw = rec.strip()
        if not raw:
            continue
        key = raw.lower()
        if key in hosts:
            continue
        entry = counts.setdefault(key, {"name": raw, "count": 0})
        entry["count"] += 1
    items = sorted(
        counts.values(),
        key=lambda e: (-e["count"], _fr_sort_key(e["name"])),
    )
    # M26-19 : dédoublonnement des slugs (parité TS `uniqueSlug`).
    used: set[str] = set()
    out: list[TopGuest] = []
    for e in items[:limit]:
        out.append(
            TopGuest(name=e["name"], slug=_unique_slug(e["name"], used), count=e["count"])
        )
    return out


def unique_slug(name: str, used: set[str]) -> str:
    """Pendant Python de `slug.ts::uniqueSlug` (M26-19).

    B-NIT-6 — exposé publiquement (sans préfixe `_`) car réutilisé par
    `compute_top_guests`. L'alias historique ``_unique_slug`` est
    conservé pour rétro-compat des imports existants.
    """
    return _unique_slug_impl(name, used)


def _unique_slug_impl(name: str, used: set[str]) -> str:
    root = _slugify(name)
    if root not in used:
        used.add(root)
        return root
    n = 2
    while f"{root}-{n}" in used:
        n += 1
    out = f"{root}-{n}"
    used.add(out)
    return out


#: Alias historique : conserve les imports existants `_unique_slug`.
_unique_slug = _unique_slug_impl


def compute_top_works(
    items: list[Mapping[str, Any]],
    mentions: list[Mapping[str, Any]],
    limit: int = 10,
) -> list[TopWork]:
    counts: dict[str, int] = {}
    for m in public_mentions(mentions):
        iid = str(m["itemId"])
        counts[iid] = counts.get(iid, 0) + 1
    by_id = {str(it["id"]): it for it in items}
    out: list[TopWork] = []
    for iid, n in counts.items():
        it = by_id.get(iid)
        if not it:
            continue
        types = it.get("types") or []
        primary = str(types[0]) if types else "autre"
        out.append(
            TopWork(
                id=iid,
                title=str(it["title"]),
                type=primary,
                mentionsCount=n,
            )
        )
    out.sort(key=lambda w: (-w.mentionsCount, _fr_sort_key(w.title)))
    return out[:limit]


def compute_type_distribution(
    items: list[Mapping[str, Any]],
    mentions: list[Mapping[str, Any]],
) -> dict[str, int]:
    mentioned = {str(m["itemId"]) for m in public_mentions(mentions)}
    dist: dict[str, int] = {}
    for it in items:
        if str(it["id"]) not in mentioned:
            continue
        types = it.get("types") or []
        key = str(types[0]) if types else "autre"
        dist[key] = dist.get(key, 0) + 1
    # L26-27 : tri secondaire count DESC puis alpha (parité TS).
    return dict(
        sorted(dist.items(), key=lambda kv: (-kv[1], _fr_sort_key(kv[0])))
    )


def compute_monthly_episodes(
    episodes: list[Mapping[str, Any]],
) -> list[MonthlyBucket]:
    """Agrège par mois ISO `YYYY-MM` et **remplit les mois manquants** entre
    min et max avec ``count=0`` (R-P3-30 — parité TS).
    """
    counts: dict[str, int] = {}
    for e in episodes:
        m = _month_key(e.get("date"))
        if not m:
            continue
        counts[m] = counts.get(m, 0) + 1
    if not counts:
        return []
    present = sorted(counts)
    first, last = present[0], present[-1]
    y, mo = int(first[:4]), int(first[5:7])
    y_last, mo_last = int(last[:4]), int(last[5:7])
    out: list[MonthlyBucket] = []
    while (y, mo) <= (y_last, mo_last):
        key = f"{y:04d}-{mo:02d}"
        out.append(MonthlyBucket(month=key, count=counts.get(key, 0)))
        mo += 1
        if mo > 12:
            mo = 1
            y += 1
    return out


def compute_per_source(
    *,
    sources: list[Mapping[str, Any]],
    episodes: list[Mapping[str, Any]],
    mentions: list[Mapping[str, Any]],
    items: list[Mapping[str, Any]],
) -> dict[str, GlobalCounts]:
    out: dict[str, GlobalCounts] = {}
    for s in sources:
        sid = str(s["id"])
        eps = [e for e in episodes if str(e.get("sourceId")) == sid]
        mens = [
            m
            for m in mentions
            if str(((m.get("sourceRef") or {}).get("sourceId")) or "") == sid
        ]
        ids = {str(m["itemId"]) for m in mens}
        its = [i for i in items if str(i["id"]) in ids]
        out[sid] = compute_global_counts(
            sources=[s], episodes=eps, mentions=mens, items=its
        )
    return out


# --- Façade -----------------------------------------------------------------


def build_snapshot(
    *,
    sources: list[Mapping[str, Any]],
    episodes: list[Mapping[str, Any]],
    mentions: list[Mapping[str, Any]],
    items: list[Mapping[str, Any]],
    source_id: str | None = None,
    top_guests_limit: int = 50,
    top_works_limit: int = 50,
    generated_at: str | None = None,
) -> StatsSnapshot:
    """Construit le snapshot agrégé (équivalent strict de `buildStatsSnapshot`).

    Si ``source_id`` est fourni, filtre toutes les collections sur cette source.
    """
    if source_id:
        sources = [s for s in sources if str(s.get("id")) == source_id]
        episodes = [
            e for e in episodes if str(e.get("sourceId")) == source_id
        ]
        mentions = [
            m
            for m in mentions
            if str((m.get("sourceRef") or {}).get("sourceId") or "") == source_id
        ]
        kept_ids = {str(m["itemId"]) for m in mentions}
        items = [i for i in items if str(i["id"]) in kept_ids]

    if generated_at is None:
        generated_at = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

    return StatsSnapshot(
        generatedAt=generated_at,
        global_=compute_global_counts(
            sources=sources, episodes=episodes, mentions=mentions, items=items
        ),
        perSource=compute_per_source(
            sources=sources, episodes=episodes, mentions=mentions, items=items
        ),
        topGuests=compute_top_guests(mentions, sources, top_guests_limit),
        topWorks=compute_top_works(items, mentions, top_works_limit),
        typeDistribution=compute_type_distribution(items, mentions),
        monthlyEpisodes=compute_monthly_episodes(episodes),
    )
